### PR TITLE
Cuda ipc deallocation method 2 (sharing counters file in producer)

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Can't figure out why it is 14% slower than #16764

### CODE REVIEW

File: torch/csrc/CudaIPCTypes.cpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+bool HaveNewRefCounter() {
+  return next_available_file != nullptr;
+}
+
+void CreateRefCounter(std::string handle, uint64_t size, at::DataPtr data_ptr) {
+  // std::cout << "Allocating new file\n";
+  if (HaveNewRefCounter()) {
+    // ERROR
```
Comment (ezyang) [inline torch/csrc/CudaIPCTypes.cpp]: You didn't actually error out in this case...

File: torch/csrc/CudaIPCTypes.cpp [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    AT_ERROR(
+        "Producer process has been terminated before all shared CUDA tensors released.");
+  }
+}
+
+void CudaIPCSentDataLimbo::collect() {
+  std::vector<CudaIPCSentData*> kept_blocks;
+  for (auto const& sd : shared_blocks_) {
+    if (sd->get() > 0) {
+      // std::cout << "Collect keeping in limbo " << sd->get() << " " <<
```
Comment (izdeby) [inline torch/csrc/CudaIPCTypes.cpp]: There are many commented pieces of code that looks like debug logs. Clean up? 
same for lines 39-40, 48-49, 54-55, 58-59, 73, 83 in this file and others.

File: torch/csrc/CudaIPCTypes.cpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  }
+  auto rc = std::shared_ptr<RefCounterFile>(
+      new RefCounterFile(handle, size, std::move(data_ptr)));
+  rc_files[handle] = rc;
+  next_available_file = rc;
+}
+
+IPCShareData GetNewRefCounter() {
+  if (!HaveNewRefCounter()) {
+    // ERROR
```
Comment (izdeby) [inline torch/csrc/CudaIPCTypes.cpp]: Error handling is missing.

File: torch/csrc/generic/StorageSharing.cpp [hunk 4]
```diff
@@ -212,27 +212,67 @@ static PyObject * THPStorage_(newSharedFd)(PyObject *_unused, PyObject *args)
...
+    {
+      int flags = TH_ALLOCATOR_MAPPED_SHAREDMEM | TH_ALLOCATOR_MAPPED_EXCLUSIVE;
+      std::string ref_counter_handle = THPStorage_(__newHandle)();
+      at::DataPtr sptr = THRefcountedMapAllocator::makeDataPtr(
+          ref_counter_handle.c_str(), flags, sizeof(int64_t) * REF_COUNTERS_FILE_SIZE, nullptr);
+      CreateRefCounter(ref_counter_handle, REF_COUNTERS_FILE_SIZE, std::move(sptr));
+    }
+
+    auto x = GetNewRefCounter();
+    // std::cout << "Getting new ref counter " << x.handler << " " << x.offset << "\n";
```
Comment (izdeby) [inline torch/csrc/generic/StorageSharing.cpp]: Clean up?

File: torch/csrc/generic/StorageSharing.cpp [hunk 5]
```diff
@@ -241,59 +281,99 @@ static PyObject * THPStorage_(shareCuda)(THPStorage *self)
...
   // devPtr = basePtr + storage_offset
   void* devPtr = basePtr.get();
   devPtr = (char*)devPtr + storage_offset_bytes;
 
+  std::string ref_counter_handle = PyBytes_AS_STRING(_ref_counter);
+  ptrdiff_t ref_counter_offset =
+      (ptrdiff_t)THPUtils_unpackLong(_ref_counter_offset);
+  static int64_t created = 0;
+      created ++;
+      if (created % 2000 == 0) {
```
Comment (izdeby) [inline torch/csrc/generic/StorageSharing.cpp]: should there be anything inside of this if clause? Same for the one on line 363

**Pull request closed without merging**